File: packages/android/crumb-ui/src/main/kotlin/dev/crumb/ui/CrumbScreenshotArtifact.kt
package dev.crumb.ui

import android.app.Activity
import android.graphics.Bitmap
import android.graphics.Canvas
import android.graphics.Color
import android.os.Handler
import android.os.HandlerThread
import android.os.Looper
import android.view.PixelCopy
import android.view.View
import android.view.ViewGroup
import android.widget.EditText
import dev.crumb.core.CrumbArtifactManifest
import dev.crumb.core.CrumbCaptureOptions
import dev.crumb.core.CrumbPrivacyOptions
import dev.crumb.core.CrumbScreenshotMaskingState
import java.io.ByteArrayOutputStream
import java.security.MessageDigest
import java.util.UUID
import java.util.WeakHashMap
import kotlin.math.max
import kotlin.math.min
import kotlin.math.sqrt

/** Registry for host-owned Android View, ComposeView, and WebView mask regions. */
object CrumbScreenshotMasking {
    private val maskedViews = WeakHashMap<View, Unit>()

    @JvmStatic
    @JvmOverloads
    fun setMasked(view: View, masked: Boolean = true) = synchronized(maskedViews) {
        if (masked) maskedViews[view] = Unit else maskedViews.remove(view)
    }

    internal fun isMasked(view: View): Boolean = synchronized(maskedViews) {
        maskedViews.containsKey(view)
    }
}

/** Marks this view's visible bounds as an opaque region in Crumb screenshots. */
fun View.maskInCrumbScreenshots(masked: Boolean = true) {
    CrumbScreenshotMasking.setMasked(this, masked)
}

internal data class CrumbScreenshotArtifact(
    val preview: Bitmap,
    val encodedBytes: ByteArray,
    val manifest: CrumbArtifactManifest,
    val maskingState: CrumbScreenshotMaskingState,
)

internal object CrumbScreenshotArtifactPipeline {
    /**
     * Copies the host window through SurfaceFlinger and performs masking, PNG encoding, and
     * hashing off the main thread. This keeps reporter presentation independent of screenshot
     * complexity while preserving the same fail-closed artifact path as synchronous capture.
     */
    fun captureAsync(
        activity: Activity,
        capture: CrumbCaptureOptions,
        privacy: CrumbPrivacyOptions,
        completion: (CrumbScreenshotArtifact?) -> Unit,
    ) {
        val root = activity.window.decorView.rootView
        if (root.width <= 0 || root.height <= 0) {
            completion(null)
            return
        }
        val plan = capturePlan(root, privacy)
        val bitmap = runCatching {
            Bitmap.createBitmap(root.width, root.height, Bitmap.Config.ARGB_8888)
        }.getOrElse {
            completion(null)
            return
        }
        val worker = HandlerThread("Crumb screenshot").apply { start() }
        val workerHandler = Handler(worker.looper)
        val mainHandler = Handler(Looper.getMainLooper())
        val finish: (CrumbScreenshotArtifact?) -> Unit = { artifact ->
            mainHandler.post { completion(artifact) }
            worker.quitSafely()
        }

        runCatching {
            PixelCopy.request(
                activity.window,
                bitmap,
                { result ->
                    val artifact = if (result == PixelCopy.SUCCESS) {
                        runCatching { createArtifact(bitmap, plan, capture) }.getOrNull()
                    } else {
                        null
                    }
                    finish(artifact)
                },
                workerHandler,
            )
        }.onFailure { finish(null) }
    }

    fun capture(
        activity: Activity,
        capture: CrumbCaptureOptions,
        privacy: CrumbPrivacyOptions,
    ): CrumbScreenshotArtifact? {
        return runCatching {
            val root = activity.window.decorView.rootView
            if (root.width <= 0 || root.height <= 0) return null

            val plan = capturePlan(root, privacy)
            val bitmap = Bitmap.createBitmap(root.width, root.height, Bitmap.Config.ARGB_8888)
            val canvas = Canvas(bitmap)
            root.draw(canvas)
            createArtifact(bitmap, plan, capture)
        }.getOrNull()
    }

    private data class MaskRegion(
        val left: Int,
        val top: Int,
        val right: Int,
        val bottom: Int,
    )

    private data class CapturePlan(
        val maskingConfigured: Boolean,
        val maskRegions: List<MaskRegion>,
    )

    private fun capturePlan(root: View, privacy: CrumbPrivacyOptions): CapturePlan {
        val maskViews = sensitiveViews(root, privacy.maskAllTextInputs)
        val rootLocation = IntArray(2).also(root::getLocationInWindow)
        val regions = maskViews.mapNotNull { view ->
            if (
                view.visibility != View.VISIBLE || view.alpha <= 0f ||
                view.width <= 0 || view.height <= 0
            ) {
                return@mapNotNull null
            }
            val location = IntArray(2).also(view::getLocationInWindow)
            val left = location[0] - rootLocation[0] - 3
            val top = location[1] - rootLocation[1] - 3
            MaskRegion(left, top, left + view.width + 6, top + view.height + 6)
        }
        return CapturePlan(
            maskingConfigured = privacy.maskScreenshotsBeforeUpload ||
                privacy.maskAllTextInputs || maskViews.isNotEmpty(),
            maskRegions = regions,
        )
    }

    private fun createArtifact(
        bitmap: Bitmap,
        plan: CapturePlan,
        capture: CrumbCaptureOptions,
    ): CrumbScreenshotArtifact? {
        plan.maskRegions.forEach { region ->
            maskOpaque(bitmap, region.left, region.top, region.right, region.bottom)
        }
        val encoded = encode(
            bitmap = bitmap,
            maximumDimension = capture.maximumScreenshotDimension,
            maximumBytes = capture.maximumScreenshotBytes,
        ) ?: return null
        val digest = MessageDigest.getInstance("SHA-256")
            .digest(encoded.bytes)
            .joinToString("") { "%02x".format(it.toInt() and 0xff) }
        val token = UUID.randomUUID().toString().replace("-", "")
        val maskingState = if (plan.maskingConfigured) {
            CrumbScreenshotMaskingState.APPLIED
        } else {
            CrumbScreenshotMaskingState.NOT_APPLICABLE
        }
        return CrumbScreenshotArtifact(
            preview = encoded.bitmap,
            encodedBytes = encoded.bytes,
            manifest = CrumbArtifactManifest(
                id = "art_$token",
                kind = "screenshot",
                mimeType = "image/png",
                byteSize = encoded.bytes.size.toLong(),
                sha256 = digest,
                redactionState = if (plan.maskingConfigured) "masked" else "not_applicable",
                uploadId = "upl_$token",
            ),
            maskingState = maskingState,
        )
    }

    private fun sensitiveViews(view: View, includeTextInputs: Boolean): List<View> {
        val matches = mutableListOf<View>()
        if (CrumbScreenshotMasking.isMasked(view) || (includeTextInputs && view is EditText)) {
            matches += view
        }
        if (view is ViewGroup) {
            for (index in 0 until view.childCount) {
                matches += sensitiveViews(view.getChildAt(index), includeTextInputs)
            }
        }
        return matches
    }

    private fun maskOpaque(
        bitmap: Bitmap,
        left: Int,
        top: Int,
        right: Int,
        bottom: Int,
    ) {
        val clippedLeft = left.coerceIn(0, bitmap.width)
        val clippedTop = top.coerceIn(0, bitmap.height)
        val clippedRight = right.coerceIn(clippedLeft, bitmap.width)
        val clippedBottom = bottom.coerceIn(clippedTop, bitmap.height)
        val width = clippedRight - clippedLeft
        if (width == 0 || clippedBottom == clippedTop) return
        val row = IntArray(width) { Color.BLACK }
        for (y in clippedTop until clippedBottom) {
            bitmap.setPixels(row, 0, width, clippedLeft, y, width, 1)
        }
    }

    private data class EncodedBitmap(val bitmap: Bitmap, val bytes: ByteArray)

    private fun encode(
        bitmap: Bitmap,
        maximumDimension: Int,
        maximumBytes: Int,
    ): EncodedBitmap? {
        var current = resized(bitmap, maximumDimension)
        repeat(10) {
            val bytes = ByteArrayOutputStream().use { stream ->
                if (!current.compress(Bitmap.CompressFormat.PNG, 100, stream)) return null
                stream.toByteArray()
            }
            if (bytes.size <= maximumBytes) return EncodedBitmap(current, bytes)

            val longest = max(current.width, current.height)
            if (current.width <= 160 || current.height <= 160) return null
            val ratio = sqrt(maximumBytes.toDouble() / bytes.size.toDouble()) * 0.9
            val nextMaximum = max(160, (longest * min(0.9, ratio)).toInt())
            if (nextMaximum >= longest) return null
            current = resized(current, nextMaximum)
        }
        return null
    }

    private fun resized(bitmap: Bitmap, maximumDimension: Int): Bitmap {
        val longest = max(bitmap.width, bitmap.height)
        if (longest <= maximumDimension) return bitmap
        val scale = maximumDimension.toFloat() / longest.toFloat()
        return Bitmap.createScaledBitmap(
            bitmap,
            max(1, (bitmap.width * scale).toInt()),
            max(1, (bitmap.height * scale).toInt()),
            true,
        )
    }
}

File: packages/ios/Sources/CrumbUI/CrumbScreenshotArtifact.swift
import CrumbCore
#if canImport(UIKit)
import CryptoKit
import ObjectiveC
import UIKit

private enum CrumbScreenshotMaskAssociation {
    nonisolated(unsafe) static var key: UInt8 = 0
}

public extension UIView {
    /// Marks this view's visible bounds as an opaque region in Crumb screenshots.
    var crumbMaskInScreenshots: Bool {
        get {
            (objc_getAssociatedObject(self, &CrumbScreenshotMaskAssociation.key) as? Bool) == true
        }
        set {
            objc_setAssociatedObject(
                self,
                &CrumbScreenshotMaskAssociation.key,
                newValue,
                .OBJC_ASSOCIATION_RETAIN_NONATOMIC
            )
        }
    }
}

package struct CrumbScreenshotArtifact {
    package let preview: UIImage
    package let encodedData: Data
    package let manifest: CrumbArtifactManifest
    package let maskingState: CrumbScreenshotMaskingState
}

@MainActor
package enum CrumbScreenshotArtifactPipeline {
    package static func capture(
        window: UIWindow,
        capture: CrumbCaptureOptions,
        privacy: CrumbPrivacyOptions
    ) -> CrumbScreenshotArtifact? {
        guard window.bounds.width > 0, window.bounds.height > 0 else { return nil }

        let maskViews = sensitiveViews(
            in: window,
            includeTextInputs: privacy.maskAllTextInputs
        )
        let maskingConfigured = privacy.maskScreenshotsBeforeUpload
            || privacy.maskAllTextInputs
            || !maskViews.isEmpty

        let format = UIGraphicsImageRendererFormat()
        format.scale = window.screen.scale
        format.opaque = true
        let rendered = UIGraphicsImageRenderer(bounds: window.bounds, format: format).image { renderer in
            window.layer.render(in: renderer.cgContext)
            renderer.cgContext.setFillColor(UIColor.black.cgColor)
            for view in maskViews where isVisible(view) {
                let frame = view.convert(view.bounds, to: window)
                    .insetBy(dx: -3, dy: -3)
                    .intersection(window.bounds)
                guard !frame.isNull, !frame.isEmpty else { continue }
                renderer.cgContext.fill(frame)
            }
        }

        guard let encoded = encode(
            rendered,
            maximumDimension: capture.maximumScreenshotDimension,
            maximumBytes: capture.maximumScreenshotBytes
        ) else { return nil }

        let digest = SHA256.hash(data: encoded.data).map { String(format: "%02x", $0) }.joined()
        let token = UUID().uuidString.replacingOccurrences(of: "-", with: "")
        let maskingState: CrumbScreenshotMaskingState = maskingConfigured ? .applied : .notApplicable
        return CrumbScreenshotArtifact(
            preview: encoded.image,
            encodedData: encoded.data,
            manifest: CrumbArtifactManifest(
                id: "art_\(token)",
                kind: "screenshot",
                mimeType: "image/png",
                byteSize: encoded.data.count,
                sha256: digest,
                redactionState: maskingConfigured ? "masked" : "not_applicable",
                uploadID: "upl_\(token)"
            ),
            maskingState: maskingState
        )
    }

    private static func sensitiveViews(
        in view: UIView,
        includeTextInputs: Bool
    ) -> [UIView] {
        var matches: [UIView] = []
        if view.crumbMaskInScreenshots
            || (includeTextInputs && (view is UITextField || view is UITextView)) {
            matches.append(view)
        }
        for child in view.subviews {
            matches.append(contentsOf: sensitiveViews(in: child, includeTextInputs: includeTextInputs))
        }
        return matches
    }

    private static func isVisible(_ view: UIView) -> Bool {
        !view.isHidden && view.alpha > 0 && view.window != nil && view.bounds.width > 0 && view.bounds.height > 0
    }

    private static func encode(
        _ image: UIImage,
        maximumDimension: Int,
        maximumBytes: Int
    ) -> (image: UIImage, data: Data)? {
        var current = resized(image, maximumDimension: maximumDimension)
        for _ in 0..<10 {
            guard let data = current.pngData() else { return nil }
            if data.count <= maximumBytes {
                return (current, data)
            }

            let pixelSize = pixels(of: current)
            guard pixelSize.width > 160, pixelSize.height > 160 else { return nil }
            let ratio = sqrt(Double(maximumBytes) / Double(data.count)) * 0.9
            let nextMaximum = max(160, Int(max(pixelSize.width, pixelSize.height) * min(0.9, ratio)))
            guard nextMaximum < Int(max(pixelSize.width, pixelSize.height)) else { return nil }
            current = resized(current, maximumDimension: nextMaximum)
        }
        return nil
    }

    private static func resized(_ image: UIImage, maximumDimension: Int) -> UIImage {
        let pixelSize = pixels(of: image)
        let longest = max(pixelSize.width, pixelSize.height)
        guard longest > CGFloat(maximumDimension) else { return image }

        let scale = CGFloat(maximumDimension) / longest
        let target = CGSize(
            width: max(1, floor(pixelSize.width * scale)),
            height: max(1, floor(pixelSize.height * scale))
        )
        let format = UIGraphicsImageRendererFormat()
        format.scale = 1
        format.opaque = true
        return UIGraphicsImageRenderer(size: target, format: format).image { _ in
            image.draw(in: CGRect(origin: .zero, size: target))
        }
    }

    private static func pixels(of image: UIImage) -> CGSize {
        if let cgImage = image.cgImage {
            return CGSize(width: cgImage.width, height: cgImage.height)
        }
        return CGSize(width: image.size.width * image.scale, height: image.size.height * image.scale)
    }
}

#if canImport(SwiftUI)
import SwiftUI

public extension SwiftUI.View {
    /// Marks this SwiftUI view's rendered bounds as an opaque region in Crumb screenshots.
    func crumbMaskInScreenshots(_ masked: Bool = true) -> some SwiftUI.View {
        overlay(CrumbScreenshotMaskRegion(masked: masked).allowsHitTesting(false))
    }
}

private struct CrumbScreenshotMaskRegion: UIViewRepresentable {
    let masked: Bool

    func makeUIView(context: Context) -> UIView {
        let view = UIView(frame: .zero)
        view.backgroundColor = .clear
        view.isAccessibilityElement = false
        view.isUserInteractionEnabled = false
        view.crumbMaskInScreenshots = masked
        return view
    }

    func updateUIView(_ uiView: UIView, context: Context) {
        uiView.crumbMaskInScreenshots = masked
    }
}
#endif
#endif
